### ISSUE TITLE
height of ComboBox listbox can exceed layoutBounds

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#673]
DemosScreenView is used in the sun and scenery-phet demo applications. It lets you choose a demo from a combo box.  With a large number of demos (like https://github.com/phetsims/scenery-phet/issues/655) the listbox can become taller than the screen's layout bounds.   Then you have to know to resize your browser window in order to see the full listbox.  

Some options:

(1) Live with it. Require users to learn how to resize their browser window to see the full listbox.

(2) Add a scrolling options to ComboBox.  Like a Java combo box, set the maximum number of items to display in the listbox.  When that max is exceeded, add scrollbars to the listbox.  This would be a non-trivial feature to add, and is not something that has been needed in sims. It's also something that I think we would want to avoid in sims, because a scrolling listbox is (imo) not a great UX and can be confusing.

(3) Add an option to ComboBox that lets you specify maxHeight (and maxWidth, while we're at it) for the listbox. The downside of this option is that the items in the listbox are going to get smaller as the list gets longer.

(4) Don't use combo box.  Come up with another way of choosing demos.

... something else?

Labeling for discussion at dev meeting.

**Comment by samreid** [comment]
I've been satisfied with (1) for now, sometimes preferring to use query parameters to select a specific component demo.  Another option is:

(5) Find another way to split up the component demos, such as grouping them and moving some to another screen (if we haven't passed our screen limit).

**Comment by pixelzoom** [comment]
> ... moving some to another screen (if we haven't passed our screen limit).

scenery-phet demo is currently at the limit of 6 screens. But that limit is a bug, not a features, see https://github.com/phetsims/joist/issues/673.

**Comment by samreid** [comment]
Dev meeting:

@pixelzoom: This probably doesn't warrant a scrollable combo box
@zepumph: a vertical carousel could work instead?
@pixelzoom: that sounds great, I'll fix it for the demo screen view.
@zepumph: you can also use a button that pops up the carousel
@pixelzoom: let's just have it always showing.  I'll volunteer to work on this for the demo harnesses. DemoScreenView will address this.

**Comment by pixelzoom** [comment]
I created `CarouselComboBox`, a combo box that pops up a carousel, and was able to reuse `ComboBoxItem`, `ComboBoxButton`, and `Carousel`.   I'm certainly open to renaming this. And for now it lives in sun/demo/, since it's use is intentionally limited to DemoScreenView. 

Tested in all Screens that extend DemosScreenView (sun, scenery-phet, bamboo, tambo, function-builder).

Over to @zepumph for review.

**Comment by pixelzoom** [comment]
Example screenshot, from scenery-phet demo's _Components_ screen:

<img width="643" alt="screenshot_862" src="https://user-images.githubusercontent.com/3046552/107713731-1eeb7980-6c89-11eb-8fc6-0591ef5b8943.png">

**Comment by zepumph** [comment]
Looks really, really nice. Well done. I personally would want to see a PageControl accompanying the Carousel (for framing how many items there are), but see that that is not built into the Carousel type, and so may be challenging to add to this ComboBox implementation. 

I see you put the type inside the `demo/` directory, which I understand to mean that you don't think this is ready for production. I think that is acceptable, though I feel the more idiomatic thing to do is to put it top level, with a NOTE or TODO in the top documentation clearly marking this type not ready for production (which you have). I like your list of potential next steps.

I think it is so very cool that you were able to reuse so many types from ComboBox, clearly your latest refactor in that type is paying dividends!


* using a disposeEmitter may be better that keeping `multilinks`. individually.
* If dispose happens in reverse order, then button should be disposed before carousel.

Overall this is great. Feel free to close.

**Comment by pixelzoom** [comment]
I'll keep this open for a little bit and mull over the possibility of adding the page controls.  @zepumph I saw that you did that in Interaction Dashboard.   If I made that change, and moved to top-level of sun, could you use it in Interaction Dashboard?

**Comment by zepumph** [comment]
Yes definitely. Though that data visualization project isn't in active development, I'd be happy to add it to help make CarouselComboBox.js more robust.


No pressure either way.

**Comment by zepumph** [comment]
In that sim, it was vital because of how many items were in the list.

**Comment by pixelzoom** [comment]
Summary of changes:

* Added PageControl
* Fixed order of subcomponent disposal
* Factored out inner class CarouselItemNode
* Moved CarouselComboBox to sun/js/

Back to @zepumph for review, and possible integration into Interaction Dashboard.

**Comment by zepumph** [comment]
Excellent looking stuff. I used it in interaction dashboard successfully. It looks much nicer than my hacked button + carousel before. 

Because CarouselComboBox doesn't support a list parent, I did some hackary with an HStrut placeholder, otherwise opening up the combobox would push all the other data down to be parallel with the bottom of the open carousel list. I spent about 10 minutes trying to quickly add list parent support to CarouselComboBox, then gave up because I was just copying so much from combo box. If this element is to ever become production-ready, I think the best way forward would be to have an interface/mixin that could share some code between ComboBox and it.

I'm ready to close this issue. Thanks for making this @pixelzoom,
